pkg/controller/shoot/shoot_quota_control.go

```diff
@@ -0,0 +1,195 @@
+)
+
+func (c *Controller) shootQuotaAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	c.shootQuotaQueue.Add(key)
+}
+
+func (c *Controller) shootQuotaDelete(obj interface{}) {
+	shoot, ok := obj.(*gardenv1beta1.Shoot)
+	if shoot == nil || !ok {
+		return
+	}
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		return
+	}
+	c.shootQuotaQueue.Done(key)
+}
+
+func (c *Controller) reconcileShootQuotaKey(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	shoot, err := c.shootLister.Shoots(namespace).Get(name)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if err := c.quotaControl.CheckQuota(shoot, key); err != nil {
+		c.shootQuotaQueue.AddAfter(key, 2*time.Minute)
+		return nil
+	}
+	c.shootQuotaQueue.AddAfter(key, time.Hour)
+	return nil
+}
+
+// QuotaControlInterface implements the control logic for quota management of Shoots. It is implemented as an interface to allow
+// for extensions that provide different semantics. Currently, there is only one implementation.
+type QuotaControlInterface interface {
+	CheckQuota(shoot *gardenv1beta1.Shoot, key string) error
+}
+
+// NewDefaultQuotaControl returns a new instance of the default implementation of QuotaControlInterface
+// which implements the semantics for controling the quota handling of Shoot resources.
+func NewDefaultQuotaControl(k8sGardenClient kubernetes.Client, k8sGardenInformers gardeninformers.Interface) QuotaControlInterface {
+	return &defaultQuotaControl{k8sGardenClient, k8sGardenInformers}
+}
+
+type defaultQuotaControl struct {
+	k8sGardenClient    kubernetes.Client
+	k8sGardenInformers gardeninformers.Interface
+}
+
+func (c *defaultQuotaControl) CheckQuota(shootObj *gardenv1beta1.Shoot, key string) error {
+	var (
+		clusterLifeTime     *int
+		quotaReferences     []v1.ObjectReference
+		shootExpirationTime time.Time
+
+		operationID = utils.GenerateRandomString(8)
+		shoot       = shootObj.DeepCopy()
+		shootLogger = logger.NewShootLogger(logger.Logger, shoot.Name, shoot.Namespace, operationID)
+	)
+
+	switch shoot.Spec.Cloud.SecretBindingRef.Kind {
+	case "PrivateSecretBinding":
+		psb, err := c.k8sGardenInformers.
+			PrivateSecretBindings().
+			Lister().
+			PrivateSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return err
+		}
+		quotaReferences = psb.Quotas
+	case "CrossSecretBinding":
+		csb, err := c.k8sGardenInformers.
+			CrossSecretBindings().
+			Lister().
+			CrossSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return err
+		}
+		quotaReferences = csb.Quotas
+	default:
+		return fmt.Errorf("Unknown Binding type %s", shoot.Spec.Cloud.SecretBindingRef.Kind)
+	}
+
+	for _, quotaRef := range quotaReferences {
+		quota, err := c.k8sGardenInformers.
```

Comment:
Can we write these statements into one line (line breakings are not used anywhere else in the code)?

Reply:
done